sklearn/manifold/t_sne.py

```diff
@@ -769,6 +892,43 @@ def _fit(self, X, skip_num_points=0):
                 neighbors_nn = neighbors_nn[:, 1:]
             P = _joint_probabilities_nn(distances, neighbors_nn,
                                         self.perplexity, self.verbose)
+        elif self.method == 'barnes_hut2':
+            if self.verbose:
+                print("[t-SNE] Computing %i nearest neighbors..." % k)
+            if self.metric == 'precomputed':
+                # Use the precomputed distances to find
+                # the k nearest neighbors and their distances
+                neighbors_nn = np.argsort(distances, axis=1)[:, :k]
+                # TODO: define distances
+            else:
+                # Find the nearest neighbors for every point
+                # TODO: argument for class knn_estimator=None
+                # TODO: assert that the knn metric is euclidean
+                knn = NearestNeighbors(algorithm='ball_tree').fit(X)
+                # LvdM uses 3 * perplexity as the number of neighbors
+                # And we add one to not count the data point itself
+                # In the event that we have very small # of points
+                # set the neighbors to n - 1
+                distances_nn, neighbors_nn = knn.kneighbors(
+                    X, n_neighbors=k + 1)
+                distances_nn = distances_nn[:, 1:]
```

Comment:
This isn't quite doing the right thing. If two instances are identical, you can't be certain which will be output first. Use `kneighbors(None)` instead.